stratum/hal/lib/common/phaldb_service.cc

```diff
@@ -0,0 +1,425 @@
+        attrs[path] = update.value().int32_val();
+        break;
+      }
+      case phal::UpdateValue::kInt64Val: {
+        attrs[path] = update.value().int64_val();
+        break;
+      }
+      case phal::UpdateValue::kUint32Val: {
+        attrs[path] = update.value().uint32_val();
+        break;
+      }
+      case phal::UpdateValue::kUint64Val: {
+        attrs[path] = update.value().uint64_val();
+        break;
+      }
+      case phal::UpdateValue::kBoolVal: {
+        attrs[path] = update.value().bool_val();
+        break;
+      }
+      case phal::UpdateValue::kStringVal: {
+        attrs[path] = update.value().string_val();
+        break;
+      }
+      case phal::UpdateValue::kBytesVal: {
+        attrs[path] = update.value().bytes_val();
+        break;
+      }
+      default: {
+        attr_res = MAKE_ERROR(ERR_INVALID_PARAM) << "Unknown value type";
+        break;
+      }
+    }
+
+    // Push status onto results stack
+    results.push_back(attr_res.status());
+
+    // If we got an error set the top level status
+    if (!attr_res.ok()) {
+        status = attr_res.status();
+    }
+  }
+
+  // Do Set if we have no errors
+  if (status.ok()) {
+    // Note: all updates are passed down to PhalDB as one Set call
+    //       so we won't get individual status on each adapter attribute
+    //       update.
+    results = {};
+
+    // Do set for all attribute pairs
+    auto adapter =
+        absl::make_unique<phal::Adapter>(phal_interface_);
+    status = adapter->Set(attrs);
+  }
+
+  return ToGrpcStatus(status, results);
+}
+
+::grpc::Status PhalDBService::Subscribe(::grpc::ServerContext* context,
+    const phal::SubscribeRequest* req,
+    ::grpc::ServerWriter<phal::SubscribeResponse>* stream) {
+
+  RETURN_IF_NOT_AUTHORIZED(auth_policy_checker_, PhalDBService,
+                           Subscribe, context);
+
+  ::util::StatusOr<phal::Path> result;
+
+  // Convert to PhalDB Path
+  switch (req->query_case()) {
+    case phal::SubscribeRequest::kStr: {
+        result = ParseQuery(req->str());
+        break;
+    }
+    case phal::SubscribeRequest::kPath: {
+        result = ToPhalDBPath(req->path());
+        break;
+    }
+    default:
+        return ::grpc::Status(
+            ::grpc::StatusCode::INVALID_ARGUMENT,
+            "Invalid query in Subscribe request.");
+  }
+
+  // Save status
+  ::util::Status status = result.status();
+
+  // If conversion worked
+  if (result.ok()) {
+    auto path = result.ConsumeValueOrDie();
+
+    // Create writer and reader channels
+    std::shared_ptr<Channel<phal::PhalDB>> channel =
+        Channel<phal::PhalDB>::Create(128);
+
+    {
+        // Lock subscriber channels
+        absl::MutexLock l(&subscriber_thread_lock_);
+
+        // Save channel to subscriber channel map
+        subscriber_channels_[pthread_self()] = channel;
+    }
+
+    auto writer =
+            ChannelWriter<phal::PhalDB>::Create(channel);
```

Comment:
unused variable?

Reply:
look on line 356, it's passed down to the Subscribe request and then used to pass messages back to this process via the channel.